File: guild_of_heroes/guild_of_heroes_core/src/settlement.cpp
#include <stdexcept>
#include "settlement.hpp"

Int
guild_of_heroes::Settlement::getSize() const {
	return size;
}

Int
guild_of_heroes::Settlement::getMaxSizeInPast() const {
	return maxSizeInPast;
}

void
guild_of_heroes::Settlement::setSize(Int newSize) {
	if (newSize < 0)
		throw std::invalid_argument("Settlement size can not be negative");
	this->size = newSize;
	if (newSize > maxSizeInPast)
		maxSizeInPast = newSize;
}

bool
guild_of_heroes::Settlement::isAbandoned() const {
	return size == 0;
}

Int
guild_of_heroes::Settlement::getRaceWeight(const std::string& race) const
{
	if(racesWeights.empty())
		throw std::logic_error("Settlement state error: races not setupped");
	if (race.empty())
		throw std::invalid_argument("Race name can not be empty");

	auto raceWeight = racesWeights.find(race);
	if (raceWeight == racesWeights.end())
		return 0;
	return raceWeight->second;
}

void
guild_of_heroes::Settlement::setRaceWeight(const std::string& race, Int weight)
{
	if (weight < 0)
		throw std::invalid_argument("Race weight can not be negative");
	if(race.empty())
		throw std::invalid_argument("Race name can not be empty");

	auto raceWeight = racesWeights.find(race);
	if (weight == 0 && raceExist(raceWeight)) {
		if (existOnlyOneRace())
			throw std::logic_error("Impossible remove last race");
		racesWeights.erase(raceWeight);
	}
	else if(weight != 0)
		racesWeights[race] = weight;
}

bool
guild_of_heroes::Settlement::existOnlyOneRace() {
	return racesWeights.size() == 1;
}

bool
guild_of_heroes::Settlement::raceExist(
	std::map<std::string, Int>::iterator& raceWeight
) {
	return raceWeight != racesWeights.end();
}

guild_of_heroes::Settlement*
guild_of_heroes::Settlement::create(const std::string& name) {
	if (name.empty())
		throw std::invalid_argument("name can not be empty");
	return new Settlement(name);
}

guild_of_heroes::Settlement::Settlement(const std::string& name)
	: Nameable(name), size(1), maxSizeInPast(1) {
}

File: GuildOfHeroes/GuildOfHeroes.Core/Settlement.cs
using GuildOfHeroes.Core.Base;
using System;
using System.Collections.Generic;
using System.Text;

namespace GuildOfHeroes.Core
{
    public class Settlement : NameableBase
    {
        public int  Size
        {
            get { return size; }
            set
            {
                if (value < 0)
                    throw new ArgumentException("Settlement size can not be negative");
                size = value;
                if (value > MaxSizeInPast)
                    MaxSizeInPast = value;
            }
        }
        public int  MaxSizeInPast { get; private set; }
        public bool IsAbandoned => Size == 0;
        public bool MonoRacial => racesWeights.Count == 1;

        public int    GetRaceWeight(string race)
        {
            if (RacesNotSetupped())
                throw new InvalidOperationException("Settlement state error: races not setupped");
            if (race.Length == 0)
                throw new ArgumentException("Race name can not be empty");

            if (RaceNotExist(race))
                return 0;
            return racesWeights[race];
        }
        public void   SetRaceWeight(string race, int weight)
        {
            if (weight < 0)
                throw new ArgumentException("Race weight can not be negative");
            if (race.Length == 0)
                throw new ArgumentException("Race name can not be empty");

            if (weight == 0 && RaceExist(race))
            {
                if (MonoRacial)
                    throw new InvalidOperationException("Impossible remove last race");
                racesTotalWeights -= racesWeights[race];
                racesWeights.Remove(race);
            }
            else if (weight != 0)
            {
                if (RaceNotExist(race))
                    racesWeights.Add(race, 0);
                racesTotalWeights -= racesWeights[race];
                racesWeights[race] = weight;
                racesTotalWeights += racesWeights[race];
            }
        }
        public bool   RaceExist(string race)
        {
            return racesWeights.ContainsKey(race);
        }
        public bool   RaceNotExist(string race)
        {
            return !racesWeights.ContainsKey(race);
        }
        public int    GetRacePercent(string race)
        {
            int weight = 0;
            if (RacesNotSetupped())
                return 0;
            if (RaceExist(race))
                weight = racesWeights[race];
            long res = 10000000L * weight / racesTotalWeights;
            if (res % 10 >= 5)
                return (int)(res / 10) + 1;
            return (int)(res / 10);
        }
        public bool   RacesNotSetupped()
        {
            return racesWeights.Count == 0;
        }
        public bool   RacesSetupped()
        {
            return racesWeights.Count > 0;
        }

        public static Settlement Create(string name)
        {
            if (name.Length == 0)
                throw new ArgumentException("name can not be empty");
            return new Settlement(name);
        }



        private Settlement(string name)
            : base(name)
        {
            MaxSizeInPast = 1;
            size = 1;
            racesWeights = new Dictionary<string, int>();
            racesTotalWeights = 0;
        }

        private int size;
        private Dictionary<string, int> racesWeights;
        private int racesTotalWeights;
    }
}
